src/lib.rs

```diff
@@ -2268,6 +2270,43 @@ pub trait Itertools : Iterator {
         v.into_iter()
     }
 
+    /// Sort the k smallest elements into a new iterator, in ascending order.
+    ///
+    /// **Note:** This consumes the entire iterator, and returns the result
+    /// as a new iterator that owns its elements.  If the input contains
+    /// less than k elements, the result is equivalent to `self.sorted()`.
+    ///
+    /// This is guaranteed to use `k * sizeof(Self::Item) + O(1)` memory
+    /// and `O(n log k)` time, with `n` the number of elements in the input.
```

Comment:
@scottmcm `O(n)` time is impossible in the general case.  Otherwise, `it.k_smallest(n)` would sort a sequence of length `n` in linear time.

Re: `BinaryHeap::from_iter(it).into_sorted_iter().take(k)`, wouldn't this have strictly-worse performance characteristics, regarding both time and space?  I need to think some more about it once caffeinated.

Reply:
It's [absolutely possible](https://en.wikipedia.org/wiki/Selection_algorithm#Unordered_partial_sorting) if this only needs to find the k smallest elements and not also sort those elements -- the fact that this is _also_ sorting those elements wasn't obvious from the title.  (And I would say, the extra `k log k` of sorting them is unfortunate if someone just wants the smallest k and _doesn't_ need them sorted.  It would be nice to be able to turn them back into a `vec::IntoIter` for the people who would be fine with heap order, like `.k_smallest(10).sum()`.)

So there are plenty of options here:

Algorithm | space | time
-----------|-----|-----------
nth element (unsorted) | n | n
partial sort | n | n + k log k
full heap | n | n + k log n
k-size heap | k | n log k

(Which does emphasize that `BinaryHeap::into_iter_sorted` should only be used for not-known-ahead-of-time `k`.)

Reply:
> It's absolutely possible if this only needs to find the k smallest elements and not also sort those

Agreed; that just seemed like a worse API to be exposing, as:
- an undefined output order, which happens to be often sorted or almost-sorted (like heap order), can be confusing for the user (I myself thought for a moment that `BinaryHeap::into_vec()` sorted the result, and was surprised by that, as the examples I was testing with happened to produce a sorted result) ;
- sorting the end result doesn't change the asymptotic complexity, for any heap-based implementation.

However, if we expose something like a fixed-size heap — and I agree it's a good idea, if only for the `.extend()` type of use-cases you suggested — users who do not require a sorted result could use that directly.

> So there are plenty of options here:

You seem to be including partition-based algorithms — AFAIK, the only way to get better time complexity than O(n log k) — and those [do not work directly for iterators]; it's always possible to first collect everything into a `Vec`, use a partition-based algorithm, then truncate the `Vec` down to size `k`, but I would expect this to be much slower in practice.

[do not work directly for iterators]: https://en.wikipedia.org/wiki/Selection_algorithm#Online_selection_algorithm